src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1566,73 +1654,127 @@ abstract void onApplicationReply(
 
-                final MqttBeginExFW beginEx = mqttBeginExRW
-                                                  .wrap(extBuffer, 0, extBuffer.capacity())
-                                                  .typeId(mqttTypeId)
-                                                  .role(r -> r.set(MqttRole.SENDER))
-                                                  .clientId("client")
-                                                  .topic(topicFilter)
-                                                  .subscriptionId(0)
-                                                  .build();
+                final MqttBeginExFW beginEx = mqttBeginExRW.wrap(extBuffer, 0, extBuffer.capacity())
+                                                           .typeId(mqttTypeId)
+                                                           .role(r -> r.set(role))
+                                                           .clientId(clientId)
+                                                           .topic(topicFilter)
+                                                           .subscriptionId(subscriptionId)
+                                                           .build();
 
                 doBegin(application, routeId, initialId, traceId, authorization, affinity, beginEx);
 
                 final int topicKey = topicKey(topicFilter);
-                final MutableInteger value = activePublishers.computeIfAbsent(topicKey, key -> new MutableInteger());
+                final MutableInteger value = activeStreams.computeIfAbsent(topicKey, key -> new MutableInteger());
                 value.value++;
+
+                if (initialSlot == NO_SLOT)
+                {
+                    initialSlot = bufferPool.acquire(initialId);
+                }
+
+                if (initialSlot == NO_SLOT)
+                {
+                    cleanup(traceId, authorization);
+                }
             }
 
             private void doApplicationData(
                 long traceId,
                 long authorization,
+                MqttRole role,
                 OctetsFW payload,
                 Flyweight extension)
             {
                 assert MqttState.initialOpening(state);
 
-                refreshPublishTimeout();
-
-                doData(application, routeId, initialId, traceId, authorization, 0L, payload.sizeof(),
-                       payload.buffer(), payload.offset(), payload.sizeof(), extension);
+                switch (role)
+                {
+                case SENDER:
+                    refreshPublishTimeout();
+                    doData(application, routeId, initialId, traceId, authorization, 0L, payload.sizeof(),
+                        payload.buffer(), payload.offset(), payload.sizeof(), extension);
+                    break;
+                case RECEIVER:
+                default:
+                    break;
+                }
+                roles.add(role);
```

Comment:
Suggest we avoid this side effect and instead assert that we have `SENDER` role to be able to send (publish) `DATA` frames.